### PR TITLE
Create the 'Tables' data type

### PR DESCRIPTION
**Pull request by int-index**
This data type encapsulates the result of processing the `Grammar`.

In addition to `ActionTable` and `GotoTable`, it also includes intermediate results such as `Lr0Items`, which are needed for debug printing and info table generation.

The `find_redundancies` function has been moved from `Main.lhs` to `Tabular.lhs` almost verbatim (use `git show --color-moved=blocks` to confirm).

### CODE REVIEW

**Comment by Ericson2314** [comment]
I definitely like `genTables` over `runTabular`, but do we want to go from the function param to the `glr` boolean? I thought `happy-tabular` was supposed to be somewhat agnostic to the parsing strategy --- must maybe that's a impossible/nonsensical ideal so the `glr` bool is fine?

**Comment by int-index** [comment]
Sure, I can revert that bit.

**Comment by Ericson2314** [comment]
Thanks. I took a look at https://github.com/simonmar/happy/pull/200 to confirm and indeed the GLR stuff was pretty well separated in the end, so that goals does look attainable.

**Comment by int-index** [comment]
Well, #200 uses a `Bool` flag too, I used it as a reference.

**Comment by Ericson2314** [comment]
@int-index Ah I see. Yes, you were very right to demand we make the diffs readable. These things slipped past me.

**Comment by Ericson2314** [comment]
Also, btw, I am rebasing #200 on master and then this to queue up the next steps, so you can just worry about this for the moment.

File: src/Tabular.lhs [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+>           conflicts   = {-# SCC "Conflict" #-} (countConflicts actionTable)
+>           redundancies = find_redundancies select_reductions g actionTable
+>       in Tables { lr0items, la_spont, la_prop, lookaheads, lr1items,
+>                   gotoTable, actionTable, conflicts, redundancies }
+
+-----------------------------------------------------------------------------
+Find unused rules and tokens
+
+> find_redundancies
+>        :: (LRAction -> [Int]) -> Grammar -> ActionTable -> ([Int], [String])
```
Comment (Ericson2314) [inline src/Tabular.lhs:53]:
```suggestion
>        :: SelectReductions -> Grammar -> ActionTable -> ([Int], [String])
```

Comment (int-index) [reply src/Tabular.lhs:53]: Yeah, I couldn’t decide whether to use the type synonym or to keep the type signature verbatim when copying.

Alright, let’s change this one :-)

**Review by Ericson2314 — APPROVED** [review]

**Comment by knothed** [comment]
Nice! So do we want to merge #200 on top of this?

**Comment by int-index** [comment]
Definitely some parts of #200, since it introduces the `.cabal` files and so on.

**Comment by knothed** [comment]
Okay, will you and @Ericson2314 do this?
I would then after #200 is merged follow up with the next MR concerning the frontend and backend.

Also, I‘d like to make some more grammar-related types derive `Show`, and I’d appreciate it if `Grammar` would contain `hd` and `tl` fields so we wouldn’t always have to pass them separately. Can we integrate this in #200 or should this be a follow-up MR?

**Comment by Ericson2314** [comment]
@knothed see what I wrote in https://github.com/simonmar/happy/pull/200#issuecomment-932634187. I am adding a little checklist to that PR.

**Pull request merged**